<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Base Footprint -->
  <!-- This frame is always at the fround, represents robot position and orientation in 2d -->
  <link name="base_footprint" />

  <link name="sphere_center" />

  <!-- Base Link -->
  <link name="base_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.05" radius="0.09" />
      </geometry>
      <material name="blue"/>
    </visual>
    <xacro:inertial_cylinder mass="0.5" length="0.05" radius="0.09">
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </xacro:inertial_cylinder>
  </link>

  <!-- Counterweight -->
  <link name="counterweight">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07" />
      </geometry>
      <material name="black"/>
    </visual>
    <xacro:inertial_cylinder mass="1.0" length="0.04" radius="0.05">
      <origin xyz="0 0 0" rpy="0 0 0" />
    </xacro:inertial_cylinder>
  </link>

  <!-- Left Wheel -->
  <link name="left_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
      <geometry>
        <mesh filename="file:///$(find bb8_description)/meshes/wheel.stl" />
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
      <geometry>
        <mesh filename="file:///$(find bb8_description)/meshes/wheel.stl" />
      </geometry>
    </collision>
    <xacro:inertial_sphere mass="0.05" radius="0.076">
      <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
    </xacro:inertial_sphere>
  </link>

  <!-- Right Wheel -->
  <link name="right_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
      <geometry>
        <mesh filename="file:///$(find bb8_description)/meshes/wheel.stl" />
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
      <geometry>
        <mesh filename="file:///$(find bb8_description)/meshes/wheel.stl" />
      </geometry>
    </collision>
    <xacro:inertial_sphere mass="0.05" radius="0.076">
      <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
    </xacro:inertial_sphere>
  </link>

  <!-- Joints -->
  <joint name="base_footprint_joint" type="fixed">
    <parent link="base_footprint" />
    <child link="sphere_center" />
    <origin xyz="0 0 ${sphere_radius_o}" rpy="0 0 0"/>
  </joint>

  <joint name="base_joint" type="fixed">
    <parent link="sphere_center" />
    <child link="base_link" />
    <origin xyz="0 0 -0.1458" rpy="0 0 0"/>
  </joint>

  <joint name="counterweight_joint" type="fixed">
    <parent link="base_link" />
    <child link="counterweight" />
    <origin xyz="0 0 -0.03" rpy="0 0 0"/>
  </joint>

  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel"/>
    <origin xyz="0 ${wheel_separation/2}  0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>

  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel"/>
    <origin xyz="0 -${wheel_separation/2} 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>


  <!-- Collision Tuning for Robot and Sphere -->
  <gazebo reference="left_wheel">
    <self_collide>true</self_collide>
    <surface>
      <contact>
        <ode>
          <min_depth>0.001</min_depth> 
          <margin>0.001</margin>
        </ode>
        <collide_without_contact>false</collide_without_contact>
        <stiffness>1e6</stiffness>
        <damping>2e3</damping>
      </contact>
      <friction>
        <ode>
          <mu>1.0</mu> 
          <mu2>1.0</mu2>
        </ode>
      </friction>
    </surface>
  </gazebo>

  <gazebo reference="right_wheel">
    <self_collide>true</self_collide>
    <surface>
      <contact>
        <ode>
          <min_depth>0.001</min_depth> 
          <margin>0.001</margin>
        </ode>
        <collide_without_contact>false</collide_without_contact>
        <stiffness>1e6</stiffness>
        <damping>2e3</damping>
      </contact>
      <friction>
        <ode>
          <mu>1.0</mu> 
          <mu2>1.0</mu2>
        </ode>
      </friction>
    </surface>
  </gazebo>

</robot>